controllers/datamovement_controller.go

```diff
@@ -214,40 +254,154 @@ func (r *DataMovementReconciler) Reconcile(ctx context.Context, req ctrl.Request
 	go func() {
 		var cmd *exec.Cmd
 
-		// TODO: UserId/GroupId
-
-		if isTestEnv() {
-			cmd = exec.CommandContext(ctxCancel, "sleep", "1")
+		// Allow the data movement command to be overridden via the configmapa
+		// for testing purposes
+		if dmCmd != "" {
+			cmdList := strings.Split(dmCmd, " ")
+			cmd = exec.CommandContext(ctxCancel, cmdList[0], cmdList[1:]...)
+			// Otherwise use the actual `mpirun dcp` command
 		} else {
+			// TODO: UserId/GroupId
 			cmd = exec.CommandContext(ctxCancel,
 				"mpirun",
 				"--allow-run-as-root",
-				"-np", fmt.Sprintf("%d", len(hosts)), // # TODO: Might want to adjust this if running on a rabbit node
+				"-np", fmt.Sprintf("%d", numProcesses), // # TODO: Might want to adjust this if running on a rabbit node
 				"--host", strings.Join(hosts, ","),
-				"dcp", dm.Spec.Source.Path, dm.Spec.Destination.Path)
+				// Dump progress every second to ensure progress collection has
+				// something to collect even if the interval is not 1 second.
+				"dcp", "--progress", "1", dm.Spec.Source.Path, dm.Spec.Destination.Path,
+			)
 		}
-		log.Info("Running Command", "cmd", cmd.String())
+		cmdStatus.Command = cmd.String()
+		log.Info("Running Command", "cmd", cmdStatus.Command)
+
+		// Use a MultiWriter so that we can parse the output and save the full output at the end
+		var combinedOutBuf, parseBuf bytes.Buffer
+		cmd.Stdout = io.MultiWriter(os.Stdout, &parseBuf, &combinedOutBuf)
+		cmd.Stderr = cmd.Stdout // Combine stderr/stdout
+
+		// Use channels to sync progress collection and cmd.Wait().
+		chCommandDone := make(chan bool, 1)
+		chProgressDone := make(chan bool)
+
+		// Start the data movement command
+		cmd.Start()
+
+		// While the command is running, capture and process the output. Read
+		// lines until EOF to ensure we have the latest output. Then use the
+		// last regex match to obtain the most recent progress.
+		go func() {
+			var progressInt32 *int32
+			var elapsed metav1.Duration
+			elapsed.Duration = 0
+			progressStart := metav1.NowMicro()
+
+			// Perform the actual collection and update logic
+			parseAndUpdateProgress := func() {
+				progressOutput := ""
+
+				// Read all lines of output until EOF
```

Comment:
Is this parsing the whole buffer every call?

Reply:
It will read whatever is out there until there's nothing left (EOF). So if --progress 1 is outputting every second and our collection interval is 5s, this should read 5 lines output until EOF.